src/event_handler.py

```diff
@@ -14,7 +19,23 @@ def handle_event(self, data, headers):
             reason = "Event not supported."
         elif data["action"] != "completed":
             reason = "We cannot do anything with this workflow state."
+        elif self.secret and not valid_payload(
+            self.secret, data, headers["X-Hub-Signature"].replace("sha1=", "")
+        ):
+            http_code = 400
+            reason = "The secret you are using on your Github webhook does not match this app's secret."
         else:
             self.client.send_trace(data["workflow_job"])
 
         return reason, http_code
+
+
+def payload_signature(secret, payload):
+    return hmac.new(
+        secret.encode("utf-8"), json.dumps(payload).encode("utf-8"), hashlib.sha1
```

Comment:
this won't necessarily match properly (loading and then dumping json could have ~slightly different encodings depending on how github chooses to encode (json has several ascii-vs-non-ascii escaping flavors)) -- github signs the exact bytes of the input so we should probably do that as well.  I use something like this in pre-commit.ci:

```python
def _github_validate_signature(request: flask.Request) -> None:
    sig = hmac.new(creds.github_webhook_secret(), digestmod='sha256')
    sig.update(request.data)
    signature = f'sha256={sig.hexdigest()}'

    if 'X-Hub-Signature-256' not in request.headers:
        flask.abort(400, description='missing X-Hub-Signature-256')

    hub_signature = request.headers['X-Hub-Signature-256']
    if not hmac.compare_digest(signature, hub_signature):
        flask.abort(400, description='invalid signature')
```